### ISSUE TITLE
Use Lambda expression when possible

### ISSUE THREAD

**Issue by sijie** [apache/bookkeeper#655]
**FEATURE REQUEST**

1. Please describe the feature you are requesting.

bookkeeper uses a lot of callbacks. 

- those callbacks can be written in lambda expressions. It would be good to clean them up.
- we should use `CompletableFuture` than customized callbacks, so it will help in simplifying the handling of multiple async operations (both in cases of multiple steps and where several operations run in parallel).

2. Indicate the importance of this issue to you (blocker, must-have, should-have, nice-to-have). Are you currently using any workarounds to address this issue?

*nice-to-have*

3. Provide any additional detail on your proposed use case for this feature.

N/A

**Comment by merlimat** [comment]
> bookkeeper uses a lot of callbacks. those callbacks can be written in lambda expressions. It would be good to clean them up. 

In addition to lambdas I would suggest to use more the `CompletableFuture`. That helps a lot in simplifying the handling of multiple async operations (both in cases of multiple steps and where several operations run in parallel)

**Comment by sijie** [comment]
@merlimat +1

**Comment by sijie** [comment]
This has been fixed by #852
